### PR TITLE
EHN: Add index parameter to to_json

### LINKED ISSUE #17394: ENH: to_json(index=False)
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd

df = pd.DataFrame([[1,2,3],[4,5,6]], columns=['a','b','c'])
df.to_json(orient='split')
# '{"columns":["a","b","c"],"index":[0,1],"data":[[1,2,3],[4,5,6]]}'
```

#### Problem description

Useless index data increases dump size. Affects only orients `'split'` and `'table'`.

#### Expected Output

```python
>>>  df.to_json(orient='split', index=False)
'{"columns":["a","b","c"],"data":[[1,2,3],[4,5,6]]}'
``` 
This resembles a bit how `.to_csv(index=False)` works.

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 2.7.13.final.0
python-bits: 64
OS: Linux
OS-release: 4.10.0-28-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.20.3
pytest: 2.6.4
pip: 9.0.1
setuptools: 36.2.5
Cython: None
numpy: 1.13.1
scipy: 0.16.0
xarray: None
IPython: 5.4.1
sphinx: None
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: None
tables: None
numexpr: 2.6.2
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: 2.6 (dt dec pq3 ext lo64)
jinja2: 2.7.3
s3fs: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by reidy-p**
- [ ] closes #17394
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Examples:
```
In [1]: df = pd.DataFrame([[1, 2, 3], [4, 5, 6]], columns=['a', 'b', 'c'])

In [2]: df.to_json(orient='split', index=True)
Out[2]: 
{"columns":["a","b","c"],"index":[0,1],"data":[[1,2,3],[4,5,6]]}

In [3]: df.to_json(orient='split', index=False)
Out[3]:
{"columns":["a","b","c"],"data":[[1,2,3],[4,5,6]]}

In [4]: df.to_json(orient='table', index=True)
Out[4]:
{"schema": {"fields":[{"name":"index","type":"integer"},{"name":"a","type":"integer"},{"name":"b","type":"integer"},{"name":"c","type":"integer"}],"primaryKey":["index"],"pandas_version":"0.20.0"}, "data": [{"index":0,"a":1,"b":2,"c":3},{"index":1,"a":4,"b":5,"c":6}]}

In [5]: df.to_json(orient='table', index=False)
Out[5]:
{"schema": {"fields":[{"name":"a","type":"integer"},{"name":"b","type":"integer"},{"name":"c","type":"integer"}],"pandas_version":"0.20.0"}, "data": [{"a":1,"b":2,"c":3},{"a":4,"b":5,"c":6}]}
```

### CODE REVIEW

File: pandas/io/json/json.py [hunk 1]
```diff
@@ -108,6 +115,19 @@ def _format_axes(self):
             raise ValueError("Series index must be unique for orient="
                              "'{orient}'".format(orient=self.orient))
 
+    def write(self):
```
Comment (jreback) [inline pandas/io/json/json.py]: if you refactor the super-class to make ``.write()`` be ``._write(self, .......)`` (IOW to accept parameters), then these 2 can just be a super call

File: pandas/tests/io/json/test_pandas.py [hunk 2]
```diff
@@ -1147,3 +1147,80 @@ def test_data_frame_size_after_to_json(self):
         size_after = df.memory_usage(index=True, deep=True).sum()
 
         assert size_before == size_after
+
+    def test_index_false_to_json(self):
+        # GH 17394
+        # Testing index parameter in to_json
+        import json
```
Comment (jreback) [inline pandas/tests/io/json/test_pandas.py]: import this at the top

File: pandas/tests/io/json/test_pandas.py [hunk 3]
```diff
@@ -1147,3 +1147,80 @@ def test_data_frame_size_after_to_json(self):
...
+        s = pd.Series([1, 2, 3], name='A')
+
+        result = s.to_json(orient='split', index=False)
+        result = json.loads(result)
+
+        expected = {
+            'name': 'A',
+            'data': [1, 2, 3]
+        }
+
```
Comment (jreback) [inline pandas/tests/io/json/test_pandas.py]: can you add comments before each section of tests

**Review by jreback — CHANGES_REQUESTED** [review]
looks pretty good

File: pandas/io/json/json.py [hunk 4]
```diff
@@ -81,14 +87,15 @@ def __init__(self, obj, orient, date_format, double_precision,
...
+        self.index = index
 
         self.is_copy = None
         self._format_axes()
 
     def _format_axes(self):
         raise AbstractMethodError(self)
 
-    def write(self):
+    def _write(self):
```
Comment (jreback) [inline pandas/io/json/json.py]: not really what I meant

``.write()`` should remain in the superclass as the public interface with no arguments, ``_write(obj)`` will take an ``obj`` for example, so we dont' mutate in the subclasses (you will instead pass the revised obj in the super call)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/json/json.py (hunk 4 above)
Comment (reidy-p) [reply pandas/io/json/json.py]: So should there be a ``.write(self)`` and ``._write(self, obj)`` method in the ``Writer`` class and then a ``._write(self, obj)`` method in each of the subclasses (``SeriesWriter``, ``FrameWriter`` and ``JSONTableWriter``)? 

Then the ``._write(self, obj)`` methods in each of the subclasses call ``super()._write(obj)`` after making any necessary adjustments to ``obj``? 

And in the ``to_json()`` function at the top of the file we use ``._write(obj)`` on an instance of the relevant class?

Comment (jreback) [reply pandas/io/json/json.py]: .write() will just call ._write()

so the original api is preserved

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
looks good. rebase and ping on green.

**Comment by reidy-p** [comment]
@jreback thanks! Green now.

File: pandas/tests/io/json/test_pandas.py [hunk 5]
```diff
@@ -1147,3 +1148,84 @@ def test_data_frame_size_after_to_json(self):
         size_after = df.memory_usage(index=True, deep=True).sum()
 
         assert size_before == size_after
+
+    def test_index_false_to_json(self):
```
Comment (jreback) [inline pandas/tests/io/json/test_pandas.py]: can you parametrize this, will be a bit more readable i think.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/generic.py [hunk 6]
```diff
@@ -1671,6 +1672,13 @@ def to_json(self, path_or_buf=None, orient=None, date_format=None,
 
             .. versionadded:: 0.21.0
 
+        index : boolean, default True
+            Whether to include the index values in the JSON string. A
+            ValueError will be thrown if index is False when orient is not
```
Comment (jorisvandenbossche) [inline pandas/core/generic.py]: I would say something like "Not including the index (`index=False`) is only supported for orient 'split' and 'table'" instead

File: pandas/io/json/json.py [hunk 7]
```diff
@@ -171,11 +203,19 @@ def __init__(self, obj, orient, date_format, double_precision,
...
         self.orient = 'records'
-
-    def write(self):
-        data = super(JSONTableWriter, self).write()
+        self.index = index
+
+    def _write(self, obj, orient, double_precision, ensure_ascii,
+               date_unit, iso_dates, default_handler):
+        if not self.index:
+            obj = obj.drop('index', axis=1)
```
Comment (jorisvandenbossche) [inline pandas/io/json/json.py]: this is not robust (you can have a MultiIndex, or an index with another name, etc ..) 
You need to handle this above when `reset_index` is called (do `drop=True/False` depending on the value of `self.index`)

(and also make sure to add a test for this)

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]

File: pandas/io/json/json.py (hunk 7 above)
Comment (reidy-p) [reply pandas/io/json/json.py]: Thanks, that's a good point. I think I have it fixed now.

Comment (jorisvandenbossche) [reply pandas/io/json/json.py]: Did you push the new commits? (I don't see any change here)

Comment (reidy-p) [reply pandas/io/json/json.py]: Yeah sorry I forgot to push the new commit. I've pushed now.

File: pandas/tests/io/json/test_pandas.py [hunk 8]
```diff
@@ -1147,3 +1148,64 @@ def test_data_frame_size_after_to_json(self):
         size_after = df.memory_usage(index=True, deep=True).sum()
 
         assert size_before == size_after
+
+    @pytest.mark.parametrize('data, expected', [
```
Comment (jreback) [inline pandas/tests/io/json/test_pandas.py]: nice!

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. @jorisvandenbossche merge when satisfied.

File: pandas/io/json/json.py (hunk 7 above)
Comment (jorisvandenbossche) [reply pandas/io/json/json.py]: That looks better now!

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
@reidy-p Thanks!

**Pull request merged**